File: Services/ModerationService.cs
using System.Text.RegularExpressions;

namespace EpikChatBot.Services;

public readonly record struct ModerationResult(bool Allowed, string? Reason = null);

public sealed partial class ModerationService(IEnumerable<string> blockedWords, int floodLimit = 6, int floodWindowMs = 10000, int repeatLimit = 3)
{
  private readonly string[] _blocked = [.. blockedWords.Select(x => x.ToLowerInvariant())]; private readonly Dictionary<string, List<(long At, string Content)>> _history = []; private readonly Dictionary<string, long> _mutes = [];
  public ModerationResult Check(string room, string? senderId, string content, bool linksAllowed)
  {
    var key = $"{room}:{senderId}"; var now = Environment.TickCount64; if (_mutes.GetValueOrDefault(key) > now) return new(false, "muted");
    var lower = content.ToLowerInvariant(); if (!linksAllowed && MyRegex().IsMatch(content)) return new(false, "link");
    if (_blocked.Any(word => lower.Contains(word, StringComparison.Ordinal))) return new(false, "blocked-word");
    if (!_history.TryGetValue(key, out var entries)) _history[key] = entries = []; entries.RemoveAll(x => now - x.At > floodWindowMs);
    if (entries.Count >= floodLimit) return new(false, "flood"); var repeats = entries.Count(x => x.Content == lower); entries.Add((now, lower));
    return repeats >= repeatLimit ? new(false, "repeat") : new(true);
  }
  public void Mute(string room, string user, int durationMs) => _mutes[$"{room}:{user}"] = Environment.TickCount64 + durationMs;
  public void Unmute(string room, string user) => _mutes.Remove($"{room}:{user}");
  [GeneratedRegex(@"https?://|www\.", RegexOptions.IgnoreCase, "en-US")]
  private static partial Regex MyRegex();
}

File: legacy-node/lib/moderation-service.js
class ModerationService {
  constructor({ blockedWords = [], floodLimit = 6, floodWindowMs = 10000, repeatLimit = 3 } = {}) {
    this.blockedWords = blockedWords.map((word) => String(word).trim().toLowerCase()).filter(Boolean);
    this.floodLimit = floodLimit;
    this.floodWindowMs = floodWindowMs;
    this.repeatLimit = repeatLimit;
    this.activity = new Map();
    this.muted = new Map();
  }

  identity(roomId, senderId, senderName) {
    return `${roomId}:${String(senderId || senderName || "unknown").toLowerCase()}`;
  }

  mute(roomId, senderId, durationMs = 300000) {
    this.muted.set(`${roomId}:${String(senderId).toLowerCase()}`, Date.now() + durationMs);
  }

  unmute(roomId, senderId) {
    this.muted.delete(`${roomId}:${String(senderId).toLowerCase()}`);
  }

  check({ roomId, senderId, senderName, content, linksAllowed = true }) {
    const key = this.identity(roomId, senderId, senderName);
    const now = Date.now();
    const mutedUntil = this.muted.get(key) || 0;
    if (mutedUntil > now) return { allowed: false, reason: "muted" };
    if (mutedUntil) this.muted.delete(key);

    const normalized = String(content || "").trim().toLowerCase();
    if (this.blockedWords.some((word) => normalized.includes(word))) return { allowed: false, reason: "blocked-word" };
    if (!linksAllowed && /https?:\/\/|www\./i.test(normalized)) return { allowed: false, reason: "link" };

    const record = this.activity.get(key) || { messages: [], last: "", repeats: 0 };
    record.messages = record.messages.filter((timestamp) => now - timestamp < this.floodWindowMs);
    record.messages.push(now);
    record.repeats = record.last === normalized ? record.repeats + 1 : 1;
    record.last = normalized;
    this.activity.set(key, record);
    if (record.messages.length > this.floodLimit) return { allowed: false, reason: "flood" };
    if (record.repeats > this.repeatLimit) return { allowed: false, reason: "repeat" };
    return { allowed: true };
  }

  prune() {
    const now = Date.now();
    for (const [key, record] of this.activity) {
      if (!record.messages.some((timestamp) => now - timestamp < this.floodWindowMs)) this.activity.delete(key);
    }
    for (const [key, until] of this.muted) if (until <= now) this.muted.delete(key);
  }
}

module.exports = { ModerationService };
